### PR TITLE
Permissions in API

### PR DESCRIPTION
**Pull request by mkurek**
Connected to #1452 

API base classes/mixins:
- permissions per object (on list, on single objects, for related objects in object details)
- permissions per field (both readonly and change fields)
- custom router (hide models for which user doesn't have permissions)
- `ReversedChoiceField` - better handling `dj.choices.Choices` fields in API
- improved API urls handling

Permissions mechanism fixes:
- allow to specify permissions for M2M fields too
- fixed permissions per object with inheritance (the same check in Q was repeated multiple times, with many tautologies (empty Q())

TODO: 
- [x] move permissions parts from `ralph.api` to `ralph.permissions.api`
- [x] add unit tests
- [x] add docstrings everywhere

### CODE REVIEW

File: src/ralph/api/tests/_base.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import Permission
+from django.contrib.contenttypes.models import ContentType
+
+from ralph.tests.models import Foo
+
+
+class APIPermissionsTestMixin(object):
+    def setUp(self):
```
Comment (ar4s) [inline src/ralph/api/tests/_base.py]: convert to `setUpClass(cls)` if it's possible

File: src/ralph/data_center/urls/api.py [hunk 2]
```diff
@@ -1,12 +1,6 @@
...
+from ralph.api import router
 from ralph.data_center.views.api import DataCenterAssetViewSet
 
-router = routers.DefaultRouter()
 router.register(r'data-center-assets', DataCenterAssetViewSet)
-
-urlpatterns = [
-    url(r'^', include(router.urls)),
-]
+urlpatterns = []
```
Comment (ar4s) [inline src/ralph/data_center/urls/api.py:6]: it is necessary?

Comment (mkurek) [inline src/ralph/data_center/urls/api.py:6]: yes - see my comment here: https://github.com/allegro/ralph/pull/1717/files#diff-67e5e6c21dea1f8b8f26c4d0d7720774R7

**Comment by ar4s** [comment]
:ok:

**Comment by szok** [comment]
:ok:

**Pull request merged**